explorer/src/swap_chains/drive.rs

```diff
@@ -35,11 +35,43 @@ fn drive_sending(
     receiver: ReceiverRef,
     sender: SenderRef,
     amount: Balance,
+    pg: &mut postgres::Client,
 ) -> anyhow::Result<()> {
     swap.transition_to(SwapState::AttemptingSend {});
 
-    receiver.after_finalized(swap, amount)?;
-    let new_state = sender.send(swap, amount)?;
-    swap.transition_to(new_state);
+    let modified_row_count = insert_swap_send_attempt(&swap.id, pg)?;
+
+    if modified_row_count == 1 {
+        receiver.after_finalized(swap, amount)?;
+        let new_state = sender.send(swap, amount)?;
+        swap.transition_to(new_state);
+    }
+
     Ok(())
 }
+
+fn insert_swap_send_attempt(swap_id: &String, pg: &mut postgres::Client) -> anyhow::Result<u64> {
+    let do_insert = |pg: &mut postgres::Client| {
+        pg.execute("INSERT INTO swap_send_attempts (swap_id) VALUES ($1) ON CONFLICT DO NOTHING", &[swap_id])
+    };
+
+    let modified_row_count = do_insert(pg);
+
+    if let Err(e) = modified_row_count {
+        if let Some(db_error) = e.as_db_error() {
+            if db_error.message() == "relation \"swap_send_attempts\" does not exist" {
+                pg.execute(
+                    "CREATE TABLE swap_send_attempts (
+                        swap_id TEXT PRIMARY KEY NOT NULL
+                    )",
+                    &[],
+                )?;
+                return Ok(do_insert(pg)?);
+            }
+        }
+
+        anyhow::bail!(e);
+    }
```

Comment:
I'd do this migration (adding the table) out of band (i.e., not here, inside normal operation code, but instead on an interactive session).

Reply:
This is the same pattern we have in storage.rs. I don't disagree with the comment, but unless you're pointing to anything other than us not following the good 12-factor practices, I think it's out of scope